### PR TITLE
Allow setting udev rules from environment variables

### PR DESCRIPTION
**Pull request by CameronDiver**
To do this we allow "SyntheticKeys" in the config module, which can
point to different structures on disk.

Note that a lot of the changes are unrelated, and just enable writing tests in typescript, something which is unavailable without this PR. The sync-debug script has been made more reliable too. The only real commit that needs reviewing is https://github.com/balena-io/balena-supervisor/commit/44dd8020ba05b2369bf5d9205f8a878ab510dad2

Change-type: minor
Closes: #998
Signed-off-by: Cameron Diver <cameron@balena.io>

### CODE REVIEW

File: src/config/configJson.ts [hunk 1]
```diff
@@ -104,15 +110,26 @@ export default class ConfigJsonConfigBackend {
 				return constants.configJsonNonAtomicPath;
 			})
 			.then(configPath => {
+				console.log('this.cache here is', this.cache);
```
Comment (hedss) [inline src/config/configJson.ts]: Debug?

File: sync-debug.js [hunk 2]
```diff
@@ -19,59 +19,58 @@ const { fs } = require('mz');
...
+
 let changedFiles = [];
 let deletedFiles = [];
 
 const performLivepush = _.debounce(async (livepush, containerId, docker) => {
 	await livepush.performLivepush(changedFiles, deletedFiles);
 	changedFiles = [];
 	deletedFiles = [];
-	setupLogs(containerId, docker);
+	// setupLogs(containerId, docker);
```
Comment (hedss) [inline sync-debug.js]: Assume you've accidentally forgotten to remove post-setup position change?

File: src/config/configJson.ts [hunk 3]
```diff
@@ -153,4 +169,64 @@ export default class ConfigJsonConfigBackend {
...
+				if ('udevRules' in obj) {
+					cache.udevRules = obj.udevRules;
+				}
+			}
+		});
+		return cache;
+	}
+
+	// Modifies diskFormat
+	// Public so we can unit-test it
```
Comment (zvin) [inline src/config/configJson.ts]: `syntheticKeyToDiskFormat` is private below

File: src/config/utils.ts [hunk 4]
```diff
@@ -91,7 +92,8 @@ export function formatConfigKeys(
 	return _.pickBy(confWithoutNamespace, (_v, k) => {
 		return (
 			_.includes(allowedKeys, k) ||
-			(isConfigType && configBackend!.isBootConfigVar(k))
+			(isConfigType && configBackend!.isBootConfigVar(k)) ||
+			_.some(allowedNamespaces, n => _.startsWith(k, n))
```
Comment (zvin) [inline src/config/utils.ts]: Shouldn't this be `_.startsWith(k, n + '_')` ?

File: src/device-config.ts [hunk 5]
```diff
@@ -615,6 +646,10 @@ export class DeviceConfig {
 		conf.dtoverlay = (conf.dtoverlay as string[]).filter(s => !_.isEmpty(s));
 		return conf;
 	}
+
+	private static isPartOfValidNamespace(key: string): boolean {
+		return _.some(DeviceConfig.validNamespaces, n => _.startsWith(key, n));
```
Comment (zvin) [inline src/device-config.ts]: Same as above: shouldn't this be `n +'_'`?

**Pull request closed without merging**